mirrord/layer/src/socket/ops.rs

```diff
@@ -321,28 +319,90 @@ pub(super) fn connect(
+    let optional_ip_address = remote_address.as_socket();
 
-    let ignore_localhost = OUTGOING_IGNORE_LOCALHOST
+    let unix_streams = REMOTE_UNIX_STREAMS
         .get()
-        .copied()
         .expect("Should be set during initialization!");
 
-    if ignore_localhost && remote_address.ip().is_loopback() {
-        return Detour::Bypass(Bypass::IgnoreLocalhost(remote_address.port()));
-    }
-
-    if is_ignored_port(remote_address) || port_debug_patch(remote_address) {
-        Err(Bypass::Port(remote_address.port()))?
-    }
-
     let user_socket_info = {
         SOCKETS
             .lock()?
             .remove(&sockfd)
             .ok_or(Bypass::LocalFdNotFound(sockfd))?
     };
 
+    if let Some(ip_address) = optional_ip_address.as_ref() {
+        let ignore_localhost = OUTGOING_IGNORE_LOCALHOST
+            .get()
+            .copied()
+            .expect("Should be set during initialization!");
+
+        if ip_address.ip().is_loopback() {
+            if ignore_localhost {
+                return Detour::Bypass(Bypass::IgnoreLocalhost(ip_address.port()));
+            }
+            if let Some(res) =
+                // Iterate through sockets, if any of them has the requested port that the
+                // application is now trying to connect to - then don't forward this connection
+                // to the agent, and instead of connecting to the requested address, connect to
+                // the actual address where the application is locally listening.
+                SOCKETS.lock()?.values().find_map(|socket| {
+                    if let SocketState::Listening(Bound {
+                        requested_address,
+                        address,
+                    }) = socket.state
+                    {
+                        if requested_address.port() == ip_address.port()
+                            && socket.protocol == user_socket_info.protocol
+                        {
+                            let rawish_remote_address = SockAddr::from(address);
+                            let result = unsafe {
+                                FN_CONNECT(
+                                    sockfd,
+                                    rawish_remote_address.as_ptr(),
+                                    rawish_remote_address.len(),
+                                )
+                            };
+
+                            return Some(if result != 0 {
+                                Detour::Error(io::Error::last_os_error().into())
+                            } else {
+                                Detour::Success(result.into())
+                            });
+                        }
+                    }
+                    None
+                })
+            {
+                return res;
+            }
+        }
+
+        if is_ignored_port(ip_address) || port_debug_patch(ip_address) {
+            Err(Bypass::Port(ip_address.port()))?
```

Comment:
This line was moved from [Line 336](https://github.com/metalbear-co/mirrord/pull/1167/files#diff-f4a8d346eda22bf3db6a591bbe3a323d559a53bd93feafb1106454e24fe477aeL336) (and slightly altered), and I didn't look into this `Err(..)?` thing and just left it as is. I wrote `return Detour::...` in other places because it's a
 straightforward, simple way to return a `Detour`, now also changed this line to that :)